### ISSUE TITLE
IndexError: list index out of range

### ISSUE THREAD

**Issue by pine3825** [bwhitman/respeaker-xmos-hid#3]
I install the hidapi.
and run the python listen_and_get_position.py.
But there is index error.
Could you have and idea to fix?
-------------------------------------------------------------
jaehoon@jaehoon-UP:~/Workspace/ReSpeaker/respeaker-xmos-hid$ python listen_and_get_position.py 
ALSA lib confmisc.c:768:(parse_card) cannot find card '0'
ALSA lib conf.c:4292:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
...
...
...
ALSA lib pcm.c:2266:(snd_pcm_open_noupdate) Unknown PCM dmix
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for 4294967295, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for 4294967295, skipping unlock
Traceback (most recent call last):
  File "listen_and_get_position.py", line 110, in <module>
    main()
  File "listen_and_get_position.py", line 102, in main
    mic_gain = read_register(0x10, 1)[0]
IndexError: list index out of range
jaehoon@jaehoon-UP:~/Workspace/ReSpeaker/respeaker-xmos-hid$ ls

**Comment by JB-45** [comment]
I get exactly the same index error:
mic_gain = read_register(0x10, 1)[0]
IndexError: list index out of range
Unfortunately no ideas yet but will keep trying.  Hoping some guru will have a look....

**Comment by fredseo91** [comment]
I've been digging this issue for a few days, and i think I found a solution. 

I will explain why the console showed "list index out of range" first.
if you break down read_register() function, you can see that the user requests some data, and the device returns back the data to the user. But for some reason, _dev.read returns only first 4 bytes of the data on some devices.

So I just tried reading the data one more time, and it worked somehow.
I honestly don't really know why, but I'm sure this will fix this problem for you too

In summary, 
edit your read_register function to:
```

def read_register(register, length):
    # To read a register you send reg & 0x80, and then read it back
    # If you have blocking off the read will return none if it's too soon after

    send_data = [0, register, 0x80, length, 0, 0, 0]
    what = _dev.write(send_data)
    ret = _dev.read(len(send_data) + length)
    ret = _dev.read(len(send_data) + length)
    return ret[4:4+length] # Data comes in at the 4th byte
```
thanks

**Comment by nccfast** [comment]
is there any solution on that?

**Comment by highvoltage379** [comment]
Same here.
Firmware issue maybe ?
